### PR TITLE
Add addLargeSerialisedKeyOp

### PR DESCRIPTION
**Pull request by dcoutts**
See the haddock documentation for the function in the code.

### CODE REVIEW

File: src/Database/LSMTree/Internal/Run/Construction.hs [hunk 1]
```diff
@@ -250,6 +252,56 @@ addLargeKeyOp racc@RunAcc{..} k e =
...
+-- within a page ('PageAcc.entryWouldFitInPage') but just /happen/ to have
+-- ended up in a page on their own in an input to a merge. A page can end up
+-- with a single entry because a page boundary was needed rather than because
+-- the entry itself was too big. Furthermore, pre-serialised pages can only be
+-- used unaltered if the entry does /not/ use a 'BlobSpan', since the 'BlobSpan'
+-- typically needs to be modified. Thus the caller should use the following
+-- tests to decide if 'addLargeSerialisedKeyOp' should be used:
+--
+-- 1. The entry does not use a 'BlobSpan'.
+-- 2. The entry definately overflows onto one or more overflow pages.
```
Comment (mheinzel) [inline src/Database/LSMTree/Internal/Run/Construction.hs:277]:
```suggestion
-- 2. The entry definitely overflows onto one or more overflow pages.
```

File: src/Database/LSMTree/Internal/Run/Construction.hs [hunk 2]
```diff
@@ -250,6 +252,56 @@ addLargeKeyOp racc@RunAcc{..} k e =
...
+          RawPage.rawPageHasBlobSpanAt page 0 == 0 &&
+          RawPage.rawPageOverflowPages page > 0 &&
+          RawPage.rawPageOverflowPages page == length overflowPages) $ do
+    modifyPrimVar entryCount (+1)
+    Bloom.insert mbloom k
+
+    -- If the existing page accumulator is non-empty, we flush it, since the
+    -- new large key/op will need more than one page to itself.
+    mpagemchunkPre <- flushPageIfNonEmpty racc
+    let nOverflowPages = length overflowPages --TODO: consider using vector
```
Comment (mheinzel) [inline src/Database/LSMTree/Internal/Run/Construction.hs:300]: Given that we have the assertion, you might even use `rawPageOverflowPages` here. Not sure. :shrug:

File: src/Database/LSMTree/Internal/Run/Construction.hs [hunk 3]
```diff
@@ -250,6 +252,56 @@ addLargeKeyOp racc@RunAcc{..} k e =
...
+  -> RawPage           -- ^ The page that this key\/op is in, which must be the
+                       -- first page of a multi-page representation of a single
+                       -- key\/op /without/ a 'BlobSpan'.
+  -> [RawOverflowPage] -- ^ The overflow pages for this key\/op
+  -> ST s ([RawPage], [RawOverflowPage], [Index.Chunk])
+addLargeSerialisedKeyOp racc@RunAcc{..} k page overflowPages =
+  assert (RawPage.rawPageNumKeys page == 1 &&
+          RawPage.rawPageHasBlobSpanAt page 0 == 0 &&
+          RawPage.rawPageOverflowPages page > 0 &&
+          RawPage.rawPageOverflowPages page == length overflowPages) $ do
```
Comment (mheinzel) [inline src/Database/LSMTree/Internal/Run/Construction.hs:293]: :+1:

**Review by mheinzel — APPROVED** [review]
That should do the trick! :+1:

File: src/Database/LSMTree/Internal/Run/Construction.hs (hunk 2 above)
Comment (dcoutts) [reply src/Database/LSMTree/Internal/Run/Construction.hs:300]: Well at the moment nothing (other than the assertions) touch the page at all.

**Pull request merged**